### ISSUE TITLE
Disposable from using statement inside async stream should be disposed after reentry from yield return

### ISSUE THREAD

**Issue by BrunoZell** [dotnet/roslyn#32040]
**Version Used**: 
Visual Studio 2019 Preview 1.1
Microsoft.NETCore.App (3.0.0-preview-27122-01)

**Steps to Reproduce**:

I have a console application to use the new `IAsyncEnumerable` feature in C# 8.0. However, the following code will behave strangely when yielding out of an `using`-statement inside of the iterator block, where the disposable is not always disposed.

Example to reproduce:

```
class Program
{
    private class IdentifyableDisposable : IDisposable
    {
        private static int _currentId = 0;
        private readonly int _thisId;

        public IdentifyableDisposable()
        {
            _thisId = ++_currentId;
            Console.WriteLine($"Just created disposable #{_thisId}!");
        }

        public void Dispose() =>
            Console.WriteLine($"Just disposed disposable #{_thisId}!");
    }

    static async Task Main(string[] args)
    {
        await foreach (var item in EnumerateAsync())
        {
            Console.WriteLine(item);
        }

        Console.WriteLine("Hello World!");
        Console.ReadKey();
    }
}
```

Now compare the two different implementations of `EnumerateAsync` which both should behave equally:

**With `try-finally`**

```
public static async IAsyncEnumerable<int> EnumerateAsync()
{
    for (int i = 0; i < 10; i++)
    {
        var d = new IdentifyableDisposable();
        try
        {
            if (i % 2 == 0)
                continue; // Will dispose

            yield return i; // Will dispose
        }
        finally
        {
            d.Dispose();
        }

        await Task.Delay(TimeSpan.FromSeconds(1));
    }
}
```

**With `using`**

```
public static async IAsyncEnumerable<int> EnumerateAsync()
{
    for (int i = 0; i < 10; i++)
    {
        using (new IdentifyableDisposable())
        {
            if (i % 2 == 0)
                continue; // Will dispose

            yield return i; // Won't dispose (except after first continue)
        }

        await Task.Delay(TimeSpan.FromSeconds(1));
    }
}
```

**Expected Behavior**:
I expect all created `IdentifyableDisposable` to be disposed using the `IdentifyableDisposable.Dispose()` method and thus it should be written to the console.

**Actual Behavior**:
However, in the version with the `using`-statement all created `IdentifyableDisposable` which follow an `yield return` are not disposed, even when the `using`-block was exited after reentry into the iterator.

Console output:

**With `try-finally`**
```
Just created disposable #1!
Just disposed disposable #1!
Just created disposable #2!
Just disposed disposable #2!
1
Just created disposable #3!
Just disposed disposable #3!
Just created disposable #4!
3
Just disposed disposable #4!
Just created disposable #5!
Just disposed disposable #5!
Just created disposable #6!
5
Just disposed disposable #6!
Just created disposable #7!
Just disposed disposable #7!
Just created disposable #8!
7
Just disposed disposable #8!
Just created disposable #9!
Just disposed disposable #9!
Just created disposable #10!
9
Just disposed disposable #10!
Hello World!
```

**With `using`**
```
Just created disposable #1!
Just disposed disposable #1!
Just created disposable #2!
Just disposed disposable #2!
1
Just created disposable #3!
Just disposed disposable #3!
Just created disposable #4!
3
Just created disposable #5!
Just disposed disposable #5!
Just created disposable #6!
5
Just created disposable #7!
Just disposed disposable #7!
Just created disposable #8!
7
Just created disposable #9!
Just disposed disposable #9!
Just created disposable #10!
9
Hello World!
```

**Comment by jcouv** [comment]
This should be fixed in preview2 by https://github.com/dotnet/roslyn/pull/31527
I'll verify

**Comment by jcouv** [comment]
I verified this is fixed in recent builds of preview2 (`3.0.100-preview-010024`). I get the same output for both implementations of `EnumerateAsync`.

```C#
Just created disposable #1!
Just disposed disposable #1!
Just created disposable #2!
1
Just disposed disposable #2!
Just created disposable #3!
Just disposed disposable #3!
Just created disposable #4!
3
Just disposed disposable #4!
Just created disposable #5!
Just disposed disposable #5!
Just created disposable #6!
5
Just disposed disposable #6!
Just created disposable #7!
Just disposed disposable #7!
Just created disposable #8!
7
Just disposed disposable #8!
Just created disposable #9!
Just disposed disposable #9!
Just created disposable #10!
9
Just disposed disposable #10!
Hello World!
```

Thanks for reporting this.
I'll go ahead and close.
